File: SearchInsertPosition.java
class Solution {
    public int searchInsert(int[] nums, int target) {
        int left = 0;
        int right = nums.length;
        while (right - left > 1) {
            int mid = left + (right -left)/2;
            if (nums[mid] == target) {
                return mid;
            }
            if (nums[mid] > target)  {
                right = mid;
            } else {
                left = mid;
            }
        }
        return (nums[left] >= target ? left : right);
    }
}

File: SearchInsertPosition.scala
object Solution {
    def searchInsert(nums: Array[Int], target: Int): Int = {
        var left : Int = 0;
        var right : Int = nums.length - 1;
        while (left <= right) {
            val mid = left + (right -left)/2;
            if (nums(mid) == target) {
                return mid;
            } else if (nums(mid) > target) {
                right = mid - 1;
            } else if (nums(mid) < target) {
                left = mid + 1;
            }
        }
        return left;
    }
}
